File: 字符串/224_基本计算器.cpp
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    int calculate(string s) {
        stack<int> sign;
        sign.push(1);
        int ans = 0, num = 0, op = 1;
        for (char c: s) {
            if (c == ' ') continue;
            if (isdigit(c)) {
                num = num * 10 + (c - '0');
                continue;
            }
            ans += op * num;
            num = 0;
            switch (c) {
                case '+': op = sign.top(); break;
                case '-': op = -sign.top(); break;
                case '(': sign.push(op); break;
                case ')': sign.pop();
            }
        }
        ans += op * num;
        return ans;
    }
};

File: 字符串/224_基本计算器.py
class Solution:
    def calculate(self, s: str) -> int:
        sign = list()
        sign.append(1)
        ans, num, op = 0, 0, 1
        for c in s:
            if c == ' ':
                continue
            if c.isdigit():
                num = num * 10 + int(c)
                continue

            ans += op * num
            num = 0
            if c == '+':
                op = sign[-1]
            elif c == '-':
                op = -sign[-1]
            elif c == '(':
                sign.append(op)
            else:
                sign.pop()

        ans += op * num
        return ans
